### PR TITLE
feat(Query): Adding Reader.Decode

### PR DESCRIPTION
**Pull request by vlastahajek**
## Proposed Changes
- Added `Reader.Decode(x interface{}) errror` to convert CSV values to a struct fields or a slice
- `Reader.NextRow()` doesn't convert values 
- `Reader.Row()` returns raw CSV data

## Checklist

- [ ] CHANGELOG.md updated
- [X] Rebased/mergeable
- [X] A test has been added if appropriate
- [X] Tests pass
- [X] Commit messages are in [semantic format](https://seesparkbox.com/foundry/semantic_commit_messages)
- [X] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)

### CODE REVIEW

File: annotatedcsv/decode.go [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+//     r.Decode(&row)
+//
+// will always succeed and provide all the values in the column as strings.
+func (r *Reader) Decode(x interface{}) error {
+	t := reflect.TypeOf(x)
+	if err := r.initColumns(t, r.cols); err != nil {
+		return err
+	}
+	et := t.Elem()
+	switch et.Kind() {
```
Comment (rogpeppe) [inline annotatedcsv/decode.go:64]: I don't think you should need this switch. I think you should be able to arrange things inside `initColumns` such that after it's run, there's just a single function inside the reader that can be called to do all the decoding for a row.

e.g. (in `Reader`)

```
    // decodeRow decodes the row with the given values into v,
    // which must have type decodeType.
    decodeRow func(v reflect.Value, row []string) error
```

File: annotatedcsv/decode.go [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+
+	return nil
+}
+
+// convertColumnValue set a value from current row of given column index
+// to a struct or a slice field value
+func (r *Reader) convertColumnValue(v reflect.Value, colIndex int) error {
+	s := stringTernary(r.row[colIndex], r.cols[colIndex].Default)
+	if err := r.colSetters[colIndex](v, s); err != nil {
+		return fmt.Errorf(`cannot convert value "%s" to type "%s" at line %d: %w`, s, r.cols[colIndex].Type, r.r.Line(), err)
```
Comment (rogpeppe) [inline annotatedcsv/decode.go:151]:
```suggestion
		return fmt.Errorf(`cannot convert value %q in column of type %q to Go type %v at line %d: %w`, s, r.cols[colIndex].Type, v.Type(), r.r.Line(), err)
```

File: annotatedcsv/decode.go [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+	case canonicalTypes[durationCol]:
+		return durationKind
+	case canonicalTypes[timeColRFC], canonicalTypes[timeColRFCNano]:
+		return timeKind
+	case canonicalTypes[base64BinaryCol]:
+		if t.Elem().Kind() == reflect.Uint8 {
+			return bytesKind
+		}
+	}
+	return fieldKind(t.Kind())
```
Comment (rogpeppe) [inline annotatedcsv/decode.go:279]: This isn't quite right for interface types. We shouldn't be allowing non-empty interface types, but this logic conflates all interface types together, which means that `toInterface` will panic for non-empty interfaces.

```suggestion
        if t.Kind() == reflect.Interface && t != emptyInterfaceType {
                // Only the empty interface type is allowed.
                return reflect.Invalid
         }
	return fieldKind(t.Kind())
```

Let's add a test case for a non-empty interface too.

File: annotatedcsv/decode.go [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+
+// stringTernary returns second argument of first is empty, otherwise first
+func stringTernary(s string, d string) string {
+	if s != "" {
+		return s
+	}
+	return d
+}
+
+func init() {
```
Comment (rogpeppe) [inline annotatedcsv/decode.go:426]: FWIW I tend to prefer to avoid `init` functions like this in favour of a function initializer style:

```
var conversions = func() map[conv] fieldSetter {
    etc
}()
```

That way you never have to worry about initialization order.

File: annotatedcsv/decode_test.go [hunk 5]
```diff
@@ -0,0 +1,844 @@
...
+	s1 := &struct {
+		S []byte
+		D time.Duration
+		T time.Time
+		I int8
+		U uint8
+		B bool
+		F float32
+	}{}
+	require.Error(t, toBytes(reflect.ValueOf(s1).Elem().FieldByName("S"), "#"))
```
Comment (rogpeppe) [inline annotatedcsv/decode_test.go:600]: I'd much prefer it if all these tests were done through the public API, please.

That way the tests give us full confidence that the overall logic works, and we have freedom to refactor internally without needing to change the tests.

I'd also like to see the actual error string in the tests. Too many times I've found tests that are testing for just the existence of an error but the tests were passing with an actual error that is either unreadable or misleading or just not the error that's expected.

A lot of this looks like it would fit a table-driven test very well:

```
var singleRowDecodeTests = struct{
    testName string
    csv string
    value interface{}
    expectError string
}{{
   testName: "Simple",
   csv: `#datatype,long,double,dateTime,string
#default,,,,
,index,score,time,name
,0,3.3,2021-02-18T10:34:08.135814545Z,Thomas
`,
    value: struct{
         Score int `flux:"score"`
         Name string `flux:"name"`
    }{
        Score: 3,
        Name: "Thomas",
    },
}, {
    testName: "InvalidDuration",
    csv: `#datatype,duration
,d
,#
`,
     value: struct{
         D time.Duration `flux:"d"`
     },
     expectError: "cannot convert column d of type duration to time.Duration: invalid etc",
}}

func TestSingleRowDecode(t *testing.T) {
    for _, test := range singleRowDecodeTests {
         t.Run(test.testName, func(t *testing.T) {
             decVal := reflect.New(reflect.TypeOf(test.value))
             r := NewReader(strings.NewReader(test.csv))
             err := r.Decode(decVal.Interface())
             if test.expectError {
                 require.Error(t, err)
                 require.Equal(t, test.expectError, err.Error()
                 return
              }
              require.NoError(t, err)
              require.Equal(t, test.value, decVal.Elem().Interface())
         })
    }
}
```

You can probably cover a great deal of the functionality by populating that table, leaving other cases like reading multiple rows and different decode types for other more specific tests.

To reduce verbosity, a couple of helper functions might turn out useful (e.g. a function that returns CSV data with a single row with a given field type, field name and field value; a function that returns a struct containing a single field with a given type, value and flux tag).

File: annotatedcsv/decode_test.go [hunk 6]
```diff
@@ -0,0 +1,844 @@
...
+	s2 := &struct {
+		Index int `flux:"index"`
+	}{}
+	err = res.Decode(s2)
+	require.Error(t, err)
+	assert.Equal(t, `cannot convert value "1.0" to type "long" at line 4: strconv.ParseInt: parsing "1.0": invalid syntax`, err.Error())
+
+}
+
+// MustParseTime returns  parsed dateTime in RFC3339 and panics if it fails
```
Comment (rogpeppe) [inline annotatedcsv/decode_test.go:837]:
```suggestion
// mustParseTime parses s as an RFC3339 timestamp and panics on error.
```

File: annotatedcsv/reader.go [hunk 7]
```diff
@@ -107,6 +109,7 @@ func (r *Reader) NextSection() bool {
 		return false
 	}
 	r.cols = cols
+	r.decodeType = nil
```
Comment (rogpeppe) [inline annotatedcsv/reader.go:112]:
```suggestion
        // Zero the cached decode type because it probably won't be
        // valid for the next section.
	r.decodeType = nil
```

File: annotatedcsv/reader.go [hunk 8]
```diff
@@ -142,25 +145,14 @@ func (r *Reader) NextRow() bool {
 	return true
 }
 
-// Row returns the values in the current row of the current section.
+// Row returns the raw values in the current row of the current section.
```
Comment (rogpeppe) [inline annotatedcsv/reader.go:148]: I wonder if this should fill in default values rather than returning them empty.
WDYT?

File: annotatedcsv/decode.go [hunk 9]
```diff
@@ -0,0 +1,451 @@
...
+
+// colType is integer based representation annotated CSV types
+type colType uint
+
+const (
+	stringCol = colType(iota)
+	boolCol
+	durationCol
+	longCol
+	uLongCol
```
Comment (rogpeppe) [inline annotatedcsv/decode.go:228]: Let's make this reflect the original naming, even if it is a tad verbose:

```suggestion
	unsignedLongCol
```

File: annotatedcsv/decode.go [hunk 10]
```diff
@@ -0,0 +1,451 @@
...
+	doubleCol
+	base64BinaryCol
+	timeColRFC
+	timeColRFCNano
```
Comment (rogpeppe) [inline annotatedcsv/decode.go:232]: It would be good to be consistent about the naming of these constants and always put `Col` at the end.
```suggestion
	rfc3339Col,
	rfc3339NanoCol
```

Also, I think you should be able to make do with only a single time kind here, because the Go time package will always let you parse RFC3339Nano with RFC3339.

File: annotatedcsv/decode.go [hunk 11]
```diff
@@ -0,0 +1,451 @@
...
+// Decode decodes the current row into x, which should be
+// a pointer to a struct or a pointer to a slice.
+//
+// When it's a pointer to a struct, columns in the row are decoded into
+// appropriate fields in the struct, using the similar tag conventions
+// described by encoding/json to determine how to map
+// column names to struct fields. Tag prefix must start with "flux:":
+//
+//  type Point struct {
+//      Timestamp  time.Time `flux:"_time"`
```
Comment (rogpeppe) [inline annotatedcsv/decode.go:26]: I'm wondering if `flux` is the correct annotation here, and thinking about `csv` instead, as we're really about CSV here, and flux might have several possible output formats in the future. WDYT?

File: annotatedcsv/decode.go [hunk 12]
```diff
@@ -0,0 +1,451 @@
...
+// - duration: time.Duration
+// - dateTime: time.Time
+//
+// Any value can be decoded into a string without
+// error - the result is the value in the CSV, so
+//
+//     var row []string
+//     r.Decode(&row)
+//
+// will always succeed and provide all the values in the column as strings.
```
Comment (rogpeppe) [inline annotatedcsv/decode.go:57]:
```suggestion
// will always succeed and provide all the values in the column as strings.
//
// Similarly, any row can be decoded into []interface{}, which will convert
// all values into elements of the slice according to the rules for empty interfaces above.
//
// Slice capacity will be reused when possible.
```

**Review by rogpeppe — CHANGES_REQUESTED** [review]
Thanks very much for this! Lots of things are right, but I have a bunch of comments and suggestions as usual :)

File: annotatedcsv/decode.go [hunk 13]
```diff
@@ -0,0 +1,451 @@
...
+		if t.Elem().Kind() == reflect.Uint8 {
+			return bytesKind
+		}
+	}
+	return fieldKind(t.Kind())
+}
+
+type fieldSetter = func(v reflect.Value, s string) error
+
+type fieldTypeOf func(col Column) reflect.Type
```
[the remark this answers is not in the document]
Comment (vlastahajek) [reply annotatedcsv/decode.go:284]: Honestly, I haven't noticed a difference before. The alias came from the original sketch and the second one was added later.
I will change that.

File: annotatedcsv/decode.go [hunk 14]
```diff
@@ -0,0 +1,451 @@
...
+}
+
+// rest of supported type kinds
+const (
+	durationKind = fieldKind(255 + iota)
+	timeKind
+	bytesKind
+)
+
+// colType is integer based representation annotated CSV types
```
[the remark this answers is not in the document]
Comment (vlastahajek) [reply annotatedcsv/decode.go:220]: Ok

File: annotatedcsv/decode.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (rogpeppe) [reply annotatedcsv/decode.go:284]: tbh i'm not sure that `fieldTypeOf` is worth having - it's not used that much.

File: annotatedcsv/decode.go [hunk 15]
```diff
@@ -0,0 +1,451 @@
...
+
+// fieldKindOf determines a fieldKind for given reflect.Type
+func fieldKindOf(t reflect.Type) fieldKind {
+	switch t {
+	case canonicalTypes[durationCol]:
+		return durationKind
+	case canonicalTypes[timeColRFC], canonicalTypes[timeColRFCNano]:
+		return timeKind
+	case canonicalTypes[base64BinaryCol]:
+		if t.Elem().Kind() == reflect.Uint8 {
```
[the remark this answers is not in the document]
Comment (vlastahajek) [reply annotatedcsv/decode.go:275]: For example,  if `t` would be e.g. array of string.

[the remark this answers is not in the document]
Comment (rogpeppe) [reply annotatedcsv/decode.go:275]: That can't happen because `canonicalTypes[base64BinaryCol]` is `[]byte` which can't match `[]string` (we're doing exact type matching here, not matching by kind).

File: annotatedcsv/decode.go [hunk 16]
```diff
@@ -0,0 +1,451 @@
...
+	r.colSetters = setters
+	r.decodeType = t
+
+	return nil
+}
+
+// convertColumnValue set a value from current row of given column index
+// to a struct or a slice field value
+func (r *Reader) convertColumnValue(v reflect.Value, colIndex int) error {
+	s := stringTernary(r.row[colIndex], r.cols[colIndex].Default)
```
[the remark this answers is not in the document]
Comment (vlastahajek) [reply annotatedcsv/decode.go:149]: Ok, it was used in more places, but after a refactor it is used only once. I will inline it

File: annotatedcsv/reader.go (hunk 8 above)
Comment (vlastahajek) [reply annotatedcsv/reader.go:148]: Good idea 👍

File: annotatedcsv/decode.go (hunk 11 above)
Comment (vlastahajek) [reply annotatedcsv/decode.go:26]: `flux` sounds, indeed, specific to the Flux results. As annotatedcsv is package for generally annotated CSV data, so `csv` is better.

File: annotatedcsv/decode.go [hunk 17]
```diff
@@ -0,0 +1,451 @@
...
+// forEachField enumerates visible fields of t, finds field name and calls field function
+func forEachField(t reflect.Type, ff fieldFunc) error {
+	fields := ireflect.VisibleFields(t)
+	for _, f := range fields {
+		name := f.Name
+		if tag, ok := f.Tag.Lookup("flux"); ok {
+			if tag == "-" {
+				continue
+			}
+			name = tag
```
[the remark this answers is not in the document]
Comment (vlastahajek) [reply annotatedcsv/decode.go:167]: Ok.

File: annotatedcsv/decode.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (vlastahajek) [reply annotatedcsv/decode.go:275]: Ok, I see. So, sure,  this check is useless.

File: annotatedcsv/decode.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (vlastahajek) [reply annotatedcsv/decode.go:284]: I'm refactoring the initColumns and Decode based on your comments, this is even not needed anymore.

File: annotatedcsv/decode.go (hunk 4 above)
Comment (vlastahajek) [reply annotatedcsv/decode.go:426]: I was trying this, but as `toInterface{}` needs `conversions`, it's a cycle dependency. So I've chosen `init` as a solution

File: annotatedcsv/decode_test.go (hunk 5 above)
Comment (vlastahajek) [reply annotatedcsv/decode_test.go:600]: In fact, this is the only unit test testing by private API. It is done this way to easily cover all type combinations, even those not possible using supported column types (mostly negative tests).
As others are testing using public API by inputting CSV data, I thought, this test could just use private API to be concise as possible

File: annotatedcsv/decode.go (hunk 9 above)
Comment (vlastahajek) [reply annotatedcsv/decode.go:228]: Ok

File: annotatedcsv/decode.go (hunk 10 above)
Comment (vlastahajek) [reply annotatedcsv/decode.go:232]: Ok

File: annotatedcsv/decode.go (hunk 4 above)
Comment (rogpeppe) [reply annotatedcsv/decode.go:436]: Good point! I hadn't realised that. Maybe add a comment noting that?

File: annotatedcsv/decode_test.go (hunk 5 above)
Comment (rogpeppe) [reply annotatedcsv/decode_test.go:600]: I'd still suggest doing this via the public API. It would be great to have as much freedom to refactor as possible. I think you should be able to get it just about as terse by using a bit of reflection.

File: annotatedcsv/decode.go (hunk 3 above)
Comment (vlastahajek) [reply annotatedcsv/decode.go:289]: Ok

File: annotatedcsv/decode.go (hunk 4 above)
Comment (vlastahajek) [reply annotatedcsv/decode.go:436]: Do you mean a comment which explains why it is initialized in the `init()`?

**Comment by vlastahajek** [comment]
@rogger, I've pushed an update that addresses all remaining issues. Please, kindly check.

File: annotatedcsv/decode.go [hunk 18]
```diff
@@ -0,0 +1,463 @@
...
+func (r *Reader) Decode(x interface{}) error {
+	t := reflect.TypeOf(x)
+	if err := r.initColumns(t, r.cols); err != nil {
+		return err
+	}
+	return r.decodeRow(reflect.ValueOf(x), r.row)
+}
+
+// initColumns initializes r.colSetters for the given columns.
+// It's called once per section.
```
Comment (rogpeppe) [inline annotatedcsv/decode.go:73]: It's called every time `Decode` is called, not once per section.

```suggestion
// initColumns initializes r.colSetters for decoding the given columns
// into a value of the given type.
```

File: annotatedcsv/decode.go [hunk 19]
```diff
@@ -0,0 +1,463 @@
...
+		fieldsMap := make(map[string]reflect.StructField)
+		fields := ireflect.VisibleFields(et)
+		for _, f := range fields {
+			name := f.Name
+			if tag, ok := f.Tag.Lookup("csv"); ok {
+				switch {
+				case tag == "-":
+					continue
+				case tag == "-,":
+					tag = "-"
```
Comment (rogpeppe) [inline annotatedcsv/decode.go:97]: This case looks redundant to me - it's handled by the case below AFAICS.

File: annotatedcsv/decode.go [hunk 20]
```diff
@@ -0,0 +1,463 @@
...
+				case strings.Contains(tag, ","):
```
Comment (rogpeppe) [inline annotatedcsv/decode.go:98]: We can avoid the double search by doing:

```
if tag == "-" {
    continue
}
if i := strings.IndexByte(tag, ','); i >= 0 {
    if i != len(tag)-1 {
       return fmt.Errorf("tag attributes are not supported")
    }
    tag = tag[:i]
}
```

File: annotatedcsv/decode.go [hunk 21]
```diff
@@ -0,0 +1,463 @@
...
+				}
+				name = tag
+			}
+			fieldsMap[name] = f
+		}
+
+		for i, col := range columns {
+			f, ok := fieldsMap[col.Name]
+			if !ok {
+				continue
```
Comment (rogpeppe) [inline annotatedcsv/decode.go:114]:
```suggestion
				// The column isn't mentioned in the struct.
				continue
```

File: annotatedcsv/decode.go [hunk 22]
```diff
@@ -0,0 +1,463 @@
...
+			}
+			ftype := f.Type
+			colType, ok := columnTypes[col.Type]
+			if !ok {
+				// ignore invalid type and use string
+				colType = stringCol
+			}
+			fkind := fieldKindOf(ftype)
+			convert, ok := conversions[conv{colType, fkind}]
```
Comment (rogpeppe) [inline annotatedcsv/decode.go:123]:
```suggestion
			convert, ok := conversions[conv{colType, fieldKindOf(ftype)}]
```

File: annotatedcsv/decode.go [hunk 23]
```diff
@@ -0,0 +1,463 @@
...
+			if !ok {
+				return fmt.Errorf("cannot convert from column type %s to %v", col.Type, ftype)
+			}
+			setters[i] = func(v reflect.Value, colIndex int) error {
+				if err := r.convertColumnValue(v.FieldByIndex(f.Index), colIndex, convert); err != nil {
+					return err
+				}
```
Comment (rogpeppe) [inline annotatedcsv/decode.go:130]:
```suggestion
				return r.convertColumnValue(v.FieldByIndex(f.Index), colIndex, convert)
```

File: annotatedcsv/decode.go [hunk 24]
```diff
@@ -0,0 +1,463 @@
...
+				return nil
+			}
+		}
+		r.decodeRow = func(v reflect.Value, row []string) error {
+			v = v.Elem()
```
Comment (rogpeppe) [inline annotatedcsv/decode.go:135]:
```suggestion
			if v.IsNil() {
			    return fmt.Errorf("decode into nil %s", v.Type())
			}
			v = v.Elem()
```

or maybe a panic. Anyway, perhaps nicer to have an explicit message rather than a panic inside reflect.

File: annotatedcsv/decode.go [hunk 25]
```diff
@@ -0,0 +1,463 @@
...
+				if err := c(v, i); err != nil {
+					return err
+				}
+			}
+			return nil
+		}
+	case reflect.Slice:
+		s := et.Elem()
+		if s != stringType && s != emptyInterfaceType {
+			return fmt.Errorf("cannot decode into pointer to slice of %v", s)
```
Comment (rogpeppe) [inline annotatedcsv/decode.go:149]:
```suggestion
			return fmt.Errorf("cannot decode into *%v", s.Type())
```
?

File: annotatedcsv/decode.go [hunk 26]
```diff
@@ -0,0 +1,463 @@
...
+			if e.Cap() < c {
+				e = reflect.MakeSlice(e.Type(), c, c)
+			} else {
+				e = e.Slice(0, c)
+			}
+			for i := 0; i < c; i++ {
+				if err := r.colSetters[i](e, i); err != nil {
+					return err
+				}
+			}
```
Comment (rogpeppe) [inline annotatedcsv/decode.go:182]: Maybe slightly more efficient if the compiler isn't that clever:

```suggestion
			for _, setf := range r.colSetters {
			    if err := setf(e, i); err != nil {
			        return err
			    }
			}
```

File: annotatedcsv/decode.go [hunk 27]
```diff
@@ -0,0 +1,463 @@
...
+const (
+	stringCol = colType(iota)
+	boolCol
+	durationCol
+	longCol
+	unsignedLongCol
+	doubleCol
+	base64BinaryCol
+	rfc3339Col
+	rfc3339NanoCol
```
Comment (rogpeppe) [inline annotatedcsv/decode.go:259]: I don't think we need to distinguish between `rfc3339` and `rfc3339Nano` - I'd lose `rfc3330Nano`

File: inluxclient/query.go [hunk 28]
```diff
@@ -98,12 +98,15 @@ func (r *QueryResultReader) errorSection() error {
...
-		message, ok1 := row[0].(string)
-		reference, ok2 := row[1].(int64)
-		if !ok1 || !ok2 {
+		if row[0] == "" {
+			return errors.New("no row found in error section")
+		}
+		var e QueryError
+		if err := r.Decode(&e); err != nil {
+
 			return fmt.Errorf("unexpected column types (%T, %T) in error section", row[0], row[1])
```
Comment (rogpeppe) [inline inluxclient/query.go:107]: I'd be inclined to keep the decode error in there.
```suggestion
			return fmt.Errorf("cannot decode row in error section: %v", err)
```

**Review by rogpeppe — APPROVED** [review]
This looks brilliant, thanks! A few more minor comments and suggestions, but LGTM.

File: annotatedcsv/decode.go (hunk 20 above)
Comment (vlastahajek) [reply annotatedcsv/decode.go:98]: Good point 👍

File: annotatedcsv/decode.go (hunk 27 above)
Comment (vlastahajek) [reply annotatedcsv/decode.go:259]: ok
